corehq/apps/app_manager/decorators.py

```diff
@@ -68,7 +70,13 @@ def _safe_cached_download(request, *args, **kwargs):
             request.GET.pop('username')
 
         try:
-            request.app = get_app(domain, app_id, latest=latest, target=target)
+            latest_enabled_build = None
+            if request.GET.get('profile') and toggles.RELEASE_BUILDS_PER_PROFILE.enabled(domain):
+                latest_enabled_build = get_latest_enabled_build_for_profile(request.GET.get('profile'))
+            if latest_enabled_build:
```

Comment:
When `latest_enabled_build` returns nothing,
it means that no specific version has been enabled for this app profile, follow the default behaviour and hence it just finds the latest released version.

Reply:
I'm still not convinced that's the appropriate behavior.  What circumstances could lead to a phone requesting an update for a specific app profile, where there have been no released builds available for that profile?  My guesses are:

* The domain just turned on the feature flag, and the app was installed before
* Someone deleted the previous available builds or marked them as unavailable for this profile
* There's a bug

Here's a hypothetical scenario:
1. A project turns on this feature flag
2. They then make a new release of the app and only make it available for one specific profile
3. A phone with a _different_ profile checks for updates
4. HQ returns a response which:
  a. says that there are no available updates
  b. updates to the latest release

Updating to this latest release seems bad, because one thing we do know for sure is that the latest release is not intended for that profile.

If we think of "profile availability" as a property of individual builds, then it would make sense that:
* All builds created prior to enabling the flag are available to all profiles
* All builds which specify a profile or profiles are available only to those profiles
* All builds which _don't_ specify a profile or profiles are available to all profiles (though you could make the case that these should be available to no profiles)

Reply:
Thanks for thinking this through @esoergel. I see this being used as
1. This gets deployed and project turns on FF
2. They enable specific versions for each profile they want to restrict. I assume they would be aware of the latest versions users are using for each profile and set them accordingly. They can choose to not touch any app profiles they use for testing or otherwise which does not need restrictions.

Reply:
Right, I didn't realize that this is completely independent of the typical method of releasing a build.  One build could be "released" for a subset of profiles without actually showing up as starred on the releases page.  You're thinking of it as "releases for a specific profile" and I'm thinking of it as "profiles for a specific release".  In the former case, then I do agree that this behavior makes sense.  I still think the latter is a more intuitive mental model for app building, but it's okay for us to disagree on that.  I'm fine to drop this discussion.